### PR TITLE
TODO-003 Change task status

### PR DESCRIPTION
**Pull request by jatanolu**
Acceptance Criteria

Given a list of tasks
When an individual task is displayed
Then a checkbox is displayed to the right of the task
And the check box should default to not checked

Given a task
When the checkbox is checked
The the task is complete

Given a task
When the checkbox is unchecked
The the task is pending

Given a list of tasks
When the tasks are displayed
Then the completed tasks are shown in a separate section below the pending task list

Given a task
When the status changes between complete and pending
Then the task moves to the appropriate section

### CODE REVIEW

File: backend/views.py [hunk 1]
```diff
@@ -21,3 +21,13 @@ def addTask(request):
     except Exception as e:
         return JsonResponse({'fail': 'Task creation failed'})
     return JsonResponse({'success': True})
+
+@api_view(['PUT'])
+def updateTask(request):
+    try:
+        updatedTask = Task.objects.get(id=request.data['id'])
```
Comment (Kotauror) [inline backend/views.py:28]: more like taskToUpdate, it's not updated yet :)

File: frontend/components/Status.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import axios from "axios";
+import { useEffect, useState } from "react";
+import { ITask, Ilist } from "../src/App";
+
+function Status(props: {
```
Comment (Kotauror) [inline frontend/components/Status.tsx]: I think the name should be changed as this component does much more than rendering the `Status`.

File: frontend/components/Status.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  submited: boolean;
+  setSubmited: (Submit: boolean) => void;
+  list: ITask[];
+  setList: (list: ITask[]) => void;
+}) {
+  const [complete, setComplete] = useState<ITask[]>([]);
+  const [pending, setPending] = useState<ITask[]>([]);
+  function splitList() {
```
Comment (Kotauror) [inline frontend/components/Status.tsx]: I'd rename this one - the name `splitList` describes the programming operation, or in other words, how you achieve the goal. Naming should reflect _what_ is your goal instead of _how_ you implemented it. In this case this could be something like `divideTasksAccordingToTheirCompletedState`. Feel free to come up with your own if mine seems too wordy.

File: frontend/components/Status.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      if (props.list[x].completed) {
+        setComplete((complete) => [...complete, props.list[x]]);
+      } else {
+        setPending((pending) => [...pending, props.list[x]]);
+      }
+    }
+  }
+  async function handleChange(id: number) {
+    let updateTask = await axios.put<Ilist>("updateTask/", { id: id });
+    console.log(updateTask);
```
Comment (Kotauror) [inline frontend/components/Status.tsx]: remove console.log

File: frontend/components/Status.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            <input
+              className="checkbox"
+              type="checkbox"
+              onChange={() => handleChange(task.id)}
+              checked={task.completed}
+            />
+          </div>
+        );
+      })}
+    </>
```
Comment (Kotauror) [inline frontend/components/Status.tsx]: Lines 34-47 and 48-61 hardly differ - only by the headers and the list on #35 and #49 - please extract the logic into a new component and use it in both places.

**Review by Kotauror — CHANGES_REQUESTED** [review]
Blocked as there are no tests on the FE and the BE.

File: frontend/components/Task.tsx [hunk 6]
```diff
@@ -30,10 +28,7 @@ function Task(props: {
 
   return (
     <>
-      <p>From the task component</p>
-      {props.list.map((task) => {
-        return <p>{task.title}</p>;
-      })}
+      <Status {...props} />
```
Comment (jay-meister) [inline frontend/components/Task.tsx]: You shouldn't need to wrap this <Status /> in an empty `fragment` (`<> </>`)

File: frontend/components/Task.tsx [hunk 7]
```diff
@@ -1,19 +1,17 @@
...
   list: ITask[];
   setList: (list: ITask[]) => void;
 }) {
-  interface Ilist {
-    tasks: ITask[];
-  }
-
   const tasklist = async () => {
     let allTasks = await axios.get<Ilist>("getTasks/");
+    console.log(allTasks);
```
Comment (jay-meister) [inline frontend/components/Task.tsx]: 🔦  Another log 🪵

File: frontend/components/Status.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import axios from "axios";
+import { useEffect, useState } from "react";
+import { ITask, Ilist } from "../src/App";
```
Comment (jay-meister) [inline frontend/components/Status.tsx]: These types are lacking consistency of capitalisation: `ITask` vs `Ilist` (should be `IList`).

More generally, I don't know what the `I` represents in `ITask` and `IList`, but it seems like it _could_ be redundant, so just wanted to flag that.

File: frontend/components/Status.tsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+function Status(props: {
+  submited: boolean;
+  setSubmited: (Submit: boolean) => void;
+  list: ITask[];
```
Comment (jay-meister) [inline frontend/components/Status.tsx]: I also find the name `list` unclear. Many things in the codebase are 'lists'. A more explicit name would be `taskList` or simply `tasks`.

File: frontend/components/Status.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  setList: (list: ITask[]) => void;
+}) {
+  const [complete, setComplete] = useState<ITask[]>([]);
+  const [pending, setPending] = useState<ITask[]>([]);
+  function splitList() {
+    for (let x in props.list) {
```
Comment (jay-meister) [inline frontend/components/Status.tsx]: In line with my previous naming comment, I'd expect to see more explicit naming here:

eg: `for (let task in props.taskList) {`

File: frontend/components/Status.tsx [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      if (props.list[x].completed) {
+        setComplete((complete) => [...complete, props.list[x]]);
+      } else {
+        setPending((pending) => [...pending, props.list[x]]);
+      }
+    }
+  }
+  async function handleChange(id: number) {
+    let updateTask = await axios.put<Ilist>("updateTask/", { id: id });
```
Comment (jay-meister) [inline frontend/components/Status.tsx]: Here we create a variable: `let updateTask` but do not use it.

I'd expect you to check the response from the backend, and if the status code is 200, then we can let the user know.

File: backend/views.py [hunk 12]
```diff
@@ -21,3 +21,13 @@ def addTask(request):
     except Exception as e:
         return JsonResponse({'fail': 'Task creation failed'})
     return JsonResponse({'success': True})
+
+@api_view(['PUT'])
+def updateTask(request):
```
Comment (jay-meister) [inline backend/views.py]: I don't know python or django, but I think this path and route handler should be called something more specific to what it actually does.

eg. the path would be something like: `/toggle-task-completed` which maps to the `toggleTaskCompleted` route handler

File: frontend/components/Status.tsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  }
+  useEffect(() => {
+    setComplete([]);
+    setPending([]);
+    splitList();
+  }, [props.list]);
+  return (
+    <>
+      <h2>Pending</h2>
+      {pending.map((task) => {
```
Comment (jay-meister) [inline frontend/components/Status.tsx]: Sorry to keep banging the same drum, but I feel like naming here could be more explicit: `pendingTasks` & `completedTasks`

File: frontend/components/Status.tsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
+import axios from "axios";
+import { useEffect, useState } from "react";
+import { ITask, Ilist } from "../src/App";
+
+function Status(props: {
+  submited: boolean;
+  setSubmited: (Submit: boolean) => void;
```
Comment (jay-meister) [inline frontend/components/Status.tsx]: slight spelling error/typo: `submitted`

File: frontend/components/Status.tsx (hunk 3 above)
Comment (jay-meister) [reply frontend/components/Status.tsx]: I could be wrong as I have not been able to get it running locally, but I feel like we can remove this function altogether. 

From what I can tell, we do not need to store `completedTasks` and `pendingTasks` in state, we just want to take the `taskList` and filter depending on the value of `completed`. I would do this directly in with the JSX in the return statement.

See my code snippet the comment further down in this file.

File: frontend/components/Status.tsx (hunk 5 above)
Comment (jay-meister) [reply frontend/components/Status.tsx]: Agreed, although personally I would just extract the `TaskWithCheckbox` piece of the logic (which are identical in the above `map`s), and keep the `h2` in place.

That's just personal preference and totally up to you though.

eg:
```jsx
<h2>Pending</h2> 
{taskList.filter(task => !task.completed).map(task => <TaskWithCheckbox />)} 

<h2>Completed</h2> 
{taskList.filter(task => task.completed).map(task => <TaskWithCheckbox />)}
```

**Review by jay-meister — CHANGES_REQUESTED** [review]
Nice work @jatanolu 👏 

There are a lot of comments here, but most of mine are naming related and therefore quick changes. I also believe there's a bit of unnecessary work being done in the `Status` component.

I wasn't able to get it running so I couldn't validate my comments unfortunately.

Please message in slack if you have any questions. Once our review comments are addressed, the PR will be simplified and we can see if there are any further simplifications that can be made ✨ 


___
Further small point:

Screenshot 2023-02-21 at 10 05 30

It's good practice to keep branch names related to what they are implementing. Of course this matters more as the project & team grows.

**Comment by jatanolu** [comment]
I am till working, but I have fixed most if not all the names and I changed the logic of the task rendering. @jay-meister 's solution was more of what I was intending when I was originally working through it, so changed it in that way. I will try to have tests and anything else finished up later today.

File: backend/views.py (hunk 1 above)
Comment (jay-meister) [reply backend/views.py:28]: @Kotauror's comment was about the variable that you have defined on line 28, not the route handler name on line 26.

```py
taskToUpdate = Task.objects.get(id=request.data['id'])
```

File: frontend/components/Input.tsx [hunk 15]
```diff
@@ -3,10 +3,10 @@ import { FormEvent, useState } from "react";
 import { ITask } from "../src/App";
 
 function Input(props: {
-  submited: boolean;
+  submitted: boolean;
   setSubmited: (Submit: boolean) => void;
-  list: ITask[];
-  setList: (list: ITask[]) => void;
+  taskList: ITask[];
+  setTaskList: (taskList: ITask[]) => void;
```
Comment (jay-meister) [inline frontend/components/Input.tsx:9]: 👌

File: frontend/components/RenderTask.tsx [hunk 16]
```diff
@@ -0,0 +1,19 @@
+function RenderTask(props: {
+  id: number;
+  title: string;
+  completed: boolean;
+  func: Function;
```
Comment (jay-meister) [inline frontend/components/RenderTask.tsx:5]: - as a general rule, try not to use abbreviations when naming, it saves few characters and just introduces additional work for the reader of the code to understand
- more importantly regarding the name `func` or even `function` - in line with previous comments, name this after what it does, not what it is.. A conventional name would be: 'changeHandler'

File: frontend/components/Task.tsx [hunk 17]
```diff
@@ -1,34 +1,44 @@
...
+        .filter((task) => !task.completed)
+        .map((task) => (
+          <RenderTask {...task} func={handleChange} />
+        ))}
+      <h2>Complete</h2>
+      {props.taskList
+        .filter((task) => task.completed)
+        .map((task) => (
+          <RenderTask {...task} func={handleChange} />
+        ))}
```
Comment (jay-meister) [inline frontend/components/Task.tsx:41]: Much better 👍

File: frontend/components/Task.tsx [hunk 18]
```diff
@@ -1,34 +1,44 @@
...
   }, []);
 
   return (
     <>
-      <Status {...props} />
+      <h2>Pending</h2>
+      {props.taskList
+        .filter((task) => !task.completed)
+        .map((task) => (
+          <RenderTask {...task} func={handleChange} />
```
Comment (jay-meister) [inline frontend/components/Task.tsx:34]: `func={handleChange}`

This should be a naming red flag to you - if you change names between files, then it becomes harder to keep track.

As mentioned previously, should be: `handleChange={handleChange}`

File: frontend/components/Task.tsx [hunk 19]
```diff
@@ -1,34 +1,44 @@
...
-  setSubmited: (Submit: boolean) => void;
-  list: ITask[];
-  setList: (list: ITask[]) => void;
-}) {
-  const tasklist = async () => {
-    let allTasks = await axios.get<Ilist>("getTasks/");
-    console.log(allTasks);
-    props.setList(allTasks.data.tasks);
+function Task(props: Iprops) {
+  const tasktaskList = async () => {
```
Comment (jay-meister) [inline frontend/components/Task.tsx:7]: The casing is inconsistent, should be `taskTaskList`.

More importantly, this name is unclear. I have no idea what it does from it's name. Think about what is it doing, then think of a name which captures that.
